File: python/alphabet.py
#!/usr/bin/env python3
""" An attempt to solve the Alphabet problem on Kattis """

import sys

# This is a longest increasing sequence question.
# Strings can easily be expressed as integers via ord().  Algorithm and explanation here:
# http://www.algorithmist.com/index.php/Longest_Increasing_Subsequence
# function lis_length( a )
#     n := a.length
#     q := new Array(n)
#     for k from 0 to n:
#         max := 0;
#         for j from 0 to k, if a[k] > a[j]:
#             if q[j] > max, then set max = q[j].
#         q[k] := max + 1;
#     max := 0
#     for i from 0 to n:
#         if q[i] > max, then set max = q[i].
#     return max;


def lis_length(a):
    n = len(a)
    q = [0] * n
    for k in range(0, n):
        maximum_length = 0
        for j in range(0, k):
            if ((a[k] > a[j]) and
                    q[j] > maximum_length):
                maximum_length = q[j]
        q[k] = maximum_length + 1
    maximum_length = 0
    for i in range(0, n):
        if q[i] > maximum_length:
            maximum_length = q[i]

    return maximum_length


def main():
    data = sys.stdin.readline().rstrip()
    # Can do this on the fly in lis_length.. or just do it in one step here
    line = [ord(letter) for letter in data]
    print(26 - lis_length(line))

if __name__ == "__main__":
    main()

File: src/bin/alphabet.rs
use std::io::{self, BufRead};

// This is a longest increasing sequence question.
// Strings in rust are easily read as vectors of characters.  Algorithm and explanation here:
// http://www.algorithmist.com/index.php/Longest_Increasing_Subsequence
// function lis_length( a )
//     n := a.length
//     q := new Array(n)
//     for k from 0 to n:
//         max := 0;
//         for j from 0 to k, if a[k] > a[j]:
//             if q[j] > max, then set max = q[j].
//         q[k] := max + 1;
//     max := 0
//     for i from 0 to n:
//         if q[i] > max, then set max = q[i].
//     return max;

fn lis_length(a: Vec<u8>) -> usize {
    let n = a.len();
    let mut q = vec![0; n];
    for k in 0..n {
        let mut max = 0;
        for j in 0..k {
            if a[k] > a[j] && q[j] > max {
                max = q[j];
            };
        }
        q[k] = max + 1;
    }
    let mut max = 0;
    for i in 0..n {
        if q[i] > max {
            max = q[i];
        }
    }
    max
}

fn main() {
    let stdin = io::stdin();

    // Get the Input
    let data = stdin.lock().lines().next().unwrap().unwrap();
    let line = data.into_bytes();
    let result = 26 - lis_length(line);
    println!("{:?}", result);
}
